sklearn/datasets/_base.py

```diff
@@ -1454,28 +1464,161 @@ def _fetch_remote(remote, dirname=None, n_retries=3, delay=1):
-            "file may be corrupted.".format(file_path, checksum, remote.checksum)
-        )
     return file_path
+
+
+def _slugify(value):
+    """Derive a name that is safe to use as filename from the given string.
+
+    Adapted from
+    https://github.com/django/django/blob/master/django/utils/text.py
+
+    Convert to ASCII if 'allow_unicode' is False. Convert spaces or repeated
+    dashes to single dashes. Remove characters that aren't alphanumerics,
+    underscores, or hyphens. Convert to lowercase. Also strip leading and
+    trailing whitespace, dashes, and underscores.
+
+    Note: this version keeps "." characters unchanged contrary to the django
+    version and replace other un-authorized characters by "_".
+    """
+    value = (
+        unicodedata.normalize("NFKD", value).encode("ascii", "ignore").decode("ascii")
+    )
+    value = re.sub(r"[^.\w\s-]", "_", value.lower())
+    value = re.sub(r"_+", "_", value)
+    return re.sub(r"[-\s]+", "-", value).strip("-_")
+
+
+def _derive_folder_and_filename_from_url(url):
+    parsed_url = urlparse(url)
+    path = parsed_url.path
+    if not path:
+        path = "/"
+
+    if "/" in path:
+        base_folder, filename = path.rsplit("/", 1)
+
+    if not filename:
+        filename = "downloaded_file"
+
+    base_folder = _slugify(base_folder)
+    if base_folder:
+        base_folder = "/" + base_folder
+
+    return (
+        _slugify(parsed_url.hostname) + base_folder,
+        _slugify(filename),
+    )
+
+
+def fetch_file(
+    url, folder=None, local_filename=None, sha256=None, n_retries=3, delay=1
+):
+    """Fetch a file from the web if not already present in the local folder.
+
+    If the file already exists locally (and the SHA256 checksums match when
+    provided), the path to the local file is returned without re-downloading.
+
+    .. versionadded:: 1.6
+
+    Parameters
+    ----------
+    url : str
+        URL of the file to download.
+
+    folder : str or Path, default=None
+        Directory to save the file to. If None, the file is downloaded in a
+        folder with a name derived from the URL host name and path under
+        scikit-learn data home folder.
+
+    local_filename : str, default=None
+        Name of the file to save. If None, the filename is inferred from the
+        URL.
+
+    sha256 : str, default=None
+        SHA256 checksum of the file. If None, no checksum is verified.
+
+    n_retries : int, default=3
+        Number of retries when HTTP errors are encountered.
+
+    delay : int, default=1
+        Number of seconds between retries.
+
+    Returns
+    -------
+    file_path : Path
+        Full path of the downloaded file.
+    """
+    folder_from_url, filename_from_url = _derive_folder_and_filename_from_url(url)
+
+    if local_filename is None:
+        local_filename = filename_from_url
+
+    if folder is None:
+        folder = Path(get_data_home()) / folder_from_url
+        makedirs(folder, exist_ok=True)
```

Comment:
with the current logic, I can imagine files from very different urls/hosts ending up in the same folder, and if they have the same name, they'd override one another.

Reply:
I think this is has a very low probability on informative enough URLs, especially with the updated logic in my last commit.

If people really need to call `fetch_file` on pathological URLs, they can specify the target folder and name manually.

Reply:
yes, I like the updated logic.